### PR TITLE
Development comment part

### PR DESCRIPTION
**Pull request by Rodina222**
### Description:
Implementing the comment part of the flayspray project and its tests.
### Issues
https://github.com/codescalersinternships/Flyspray/issues/3

### CODE REVIEW

File: backend/handlers/comment.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
+package handlers
```
Comment (rawdaGastan) [inline backend/handlers/comment.go]: please refer to structure issues. Handlers should be in app dir
also name it as comment_handler

**Review by rawdaGastan — CHANGES_REQUESTED** [review]

File: backend/app/comment_handler.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	"net/http"
+
+	"gorm.io/gorm"
+
+	"github.com/codescalersinternships/Flyspray/models"
+	"github.com/gin-gonic/gin"
+)
+
+// Response is a struct that holds the ok response
+type Response struct {
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: make sure to add the wrapper

File: backend/app/comment_handler.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+// CreateComment creates a comment on a specific bug
+func (app *App) CreateComment(c *gin.Context) {
+
+	comment := models.Comment{}
+
+	if err := c.ShouldBindJSON(&comment); err != nil {
+		log.Println("failed to create the comment:", err)
+		c.JSON(http.StatusBadRequest,
+			ErrorResponse{Error: "bugID,ownerID and summary of the comment should be provided "})
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: bugID -> bug ID
ownerID -> owner ID
leave space after ,
remove the last space
errors and msgs should be user friendly as the user will deal with them, please check all

File: backend/app/comment_handler.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		return
+	}
+
+	if err := comment.Validate(); err != nil {
+		log.Println("bugID,ownerID and summary of the comment should be validated", err)
+		c.JSON(http.StatusBadRequest,
+			ErrorResponse{Error: "comment is not validated"})
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: you can tell that `the input data is invalid`

File: backend/app/comment_handler.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+		return
+	}
+
+	comment, err := app.client.CreateComment(comment)
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: I think client is misleading. db would be better

File: backend/app/comment_handler.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+	if err != nil {
+		log.Fatal(err)
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: why here fatal and others are println?

File: backend/app/comment_handler.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+	"gorm.io/gorm"
+
+	"github.com/codescalersinternships/Flyspray/models"
+	"github.com/gin-gonic/gin"
+)
+
+// Response is a struct that holds the ok response
+type Response struct {
+	Message string           `json:"message"`
+	Payload []models.Comment `json:"payload"`
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: payload shouldn' t be a slice of comments. needs to be more generic

File: backend/app/comment_handler.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+// ErrorResponse is a struct that holds the error response
+type ErrorResponse struct {
+	Error string `json:"error"`
+}
+
+// CreateComment creates a comment on a specific bug
+func (app *App) CreateComment(c *gin.Context) {
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: no need for exported handlers functions
check all

File: backend/app/comment_handler.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+		log.Println("comment ID is required")
+		c.JSON(http.StatusBadRequest, ErrorResponse{Error: "comment ID is required"})
+		return
+	}
+
+	if err := app.client.DeleteComment(id); err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			log.Println("comment is not found:", err)
+			c.JSON(http.StatusNotFound, ErrorResponse{Error: "comment is not found"})
+		} else {
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: it should be a check `if errors.Is(err, gorm.ErrRecordNotFound)` then `if err == nil`

File: backend/app/comment_handler.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+	}
+
+	if err := comment.Validate(); err != nil {
+		log.Println("bugID, ownerID, and summary of the comment should be validated", err)
+		c.JSON(http.StatusBadRequest, ErrorResponse{Error: "comment is not validated"})
+		return
+	}
+
+	comment, err := app.client.UpdateComment(comment, id)
+	if err != nil {
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go:225]: it should be a check `if errors.Is(err, gorm.ErrRecordNotFound)` then `if err == nil`

File: backend/models/comment.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+)
+
+// Comment is a struct representing comments on bugs
+type Comment struct {
+
+	// ID represents the unique id of each comment
+	ID uint `gorm:"primaryKey" json:"id"`
+
+	// OwnerID represents the unique id of the user that wrote the comment
+	OwnerID uint `json:"owner_id" validate:"required"`
```
Comment (rawdaGastan) [inline backend/models/comment.go]: UserID is better? no? and its string

File: backend/app/comment_handler.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+	c.JSON(http.StatusOK, Response{Message: "comment is deleted"})
+}
+
+// ListComments lists all the comments for a specific bug
+func (app *App) ListComments(c *gin.Context) {
+
+	bugID := c.Query("bug_id")
+
+	if bugID == "" {
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: handler can have no filters

File: backend/models/comment.go [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+	// BugID represents the unique id of the bug
+	BugID uint `json:"bug_id" validate:"required"`
+
+	// Summary represents the text of the comment
+	Summary string `json:"summary" validate:"required"`
+
+	// CreatedAt represents the time at which the comment was created
+	CreatedAt time.Time
```
Comment (rawdaGastan) [inline backend/models/comment.go]: why it doesn't have a json tag?

File: backend/models/comment.go [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+// CreateComment creates a comment on a specific bug
+func (db *DBClient) CreateComment(comment Comment) (Comment, error) {
+
+	result := db.Client.Create(&comment)
+	return comment, result.Error
+}
+
+// GetComment gets a comment on a specific bug by id
+func (db *DBClient) GetComment(id string) (Comment, error) {
```
Comment (rawdaGastan) [inline backend/models/comment.go]: id is uint? check all

File: backend/models/comment.go [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+	return comment, result.Error
+
+}
+
+// DeleteComment deletes a comment on a specific bug by id
+func (db *DBClient) DeleteComment(id string) error {
+
+	comment := Comment{}
+
+	if result := db.Client.First(&comment, id); result.Error != nil {
```
Comment (rawdaGastan) [inline backend/models/comment.go]: why you get it first?

File: backend/models/comment.go [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+	if result := db.Client.Delete(&comment, id); result.Error != nil {
+		return result.Error
+	}
+
+	return nil
+
+}
+
+// ListComments lists all the comments for a specific bug
+func (db *DBClient) ListComments(bugID string) []Comment {
```
Comment (rawdaGastan) [inline backend/models/comment.go]: handle if no filters

File: backend/models/comment.go [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+	comments := []Comment{}
+	db.Client.Where("bug_id=?", bugID).Find(&comments)
+
+	return comments
+
+}
+
+// UpdateComment updates a comment on a specific bug by id
+func (db *DBClient) UpdateComment(comment Comment, id string) (Comment, error) {
```
Comment (rawdaGastan) [inline backend/models/comment.go]: you can't take the comment as is to update. Can we update IDs?

File: backend/models/comment.go [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+
+	if result := db.Client.First(&comment, id); result.Error != nil {
```
Comment (rawdaGastan) [inline backend/models/comment.go]: why getting the comment?

File: backend/app/comment_handler.go [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+type ErrorResponse struct {
+	Error string `json:"error"`
+}
+
+// CreateComment creates a comment on a specific bug
+func (app *App) CreateComment(c *gin.Context) {
+
+	comment := models.Comment{}
+
+	if err := c.ShouldBindJSON(&comment); err != nil {
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: you should have input structs for creations and updates

**Review by rawdaGastan — CHANGES_REQUESTED** [review]

File: backend/app/comment_handler.go (hunk 6 above)
Comment (Rodina222) [reply backend/app/comment_handler.go]: I use fatal when the status is http.StatusInternalServerError as the server can't continue but with any other status code I use println.

File: backend/app/comment_handler.go [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+	"time"
+
+	"gorm.io/gorm"
+
+	"github.com/codescalersinternships/Flyspray/models"
+	"github.com/gin-gonic/gin"
+	"github.com/rs/zerolog/log"
+)
+
+type CreateCommentInput struct {
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: no need for it to be exported

File: backend/app/comment_handler.go [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+	Summary string `json:"summary" validate:"required"`
+}
+
+type updateCommentInput struct {
+	Summary string `json:"summary" validate:"required"`
+}
+
+func (app *App) createComment(c *gin.Context) (interface{}, Response) {
+
+	commentInput := CreateCommentInput{}
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: `var commentInput CreateCommentInput` is better

File: backend/app/comment_handler.go [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+	}, Created()
+
+}
+
+func (app *App) getComment(c *gin.Context) (interface{}, Response) {
+
+	idStr := c.Param("id")
+
+	if idStr == "" {
+		return nil, BadRequest(errors.New("comment ID is required"))
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: ID should be lowercase
check all

File: backend/app/comment_handler.go [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+	}
+
+	id, _ := strconv.ParseUint(idStr, 10, 64)
+
+	comment, err := app.DB.GetComment(uint(id))
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go:73]: why you used `uint`? isn't it already parsed
check all

File: backend/app/comment_handler.go [hunk 24]
```diff
@@ -0,0 +1,185 @@
...
+
+	if errors.Is(err, gorm.ErrRecordNotFound) {
+		log.Error().Err(err).Send()
+		return nil, NotFound(errors.New("comment is not found"))
+
+	}
+
+	if err != nil {
+		log.Error().Err(err).Send()
+		return nil, InternalServerError(errors.New("failed to get comment"))
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: use the `internal server msg error` for all internal server errors
check all

File: backend/app/comment_handler.go [hunk 25]
```diff
@@ -0,0 +1,185 @@
...
+
+	bugIDStr := c.Query("bug_id")
+	UserID := c.Query("user_id")
+
+	bugID, err := strconv.ParseUint(bugIDStr, 10, 64)
+	comments := app.DB.ListComments(uint(bugID), UserID)
+
+	if len(comments) == 0 {
+		log.Error().Err(err).Send()
+		return nil, NotFound(errors.New("no comments are found"))
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go]: return empty list with an Ok() response

File: backend/app/comment_handler.go [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+	}
+
+	return ResponseMsg{
+		Message: "comment is found successfully",
+		Data:    comment,
+	}, Ok()
+
+}
+
+func (app *App) deleteComment(c *gin.Context) (interface{}, Response) {
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go:93]: check delete comment access (if it's not the user)

File: backend/app/comment_handler.go [hunk 27]
```diff
@@ -0,0 +1,185 @@
...
+		return nil, NotFound(errors.New("comment is not found"))
+
+	}
+
+	if err != nil {
+		log.Error().Err(err).Send()
+		return nil, InternalServerError(errors.New("failed to get comment"))
+	}
+
+	if fmt.Sprint(userID) != fmt.Sprint(comment.UserID) {
```
Comment (rawdaGastan) [inline backend/app/comment_handler.go:214]: `you have ..`

**Review by rawdaGastan — CHANGES_REQUESTED** [review]
great work

**Comment by rawdaGastan** [comment]
add the issue in the pr description please
``` 
### Issues 
- you issue link
```

**Review by rawdaGastan — APPROVED** [review]

**Pull request merged**